Move to BYOC queues

dougbu opened the pull request:
- aspnet/AspNetCore-Internal#2033
- use conditions matching aspnet/Extensions and dotnet/Arcade
  - e.g. https://github.com/dotnet/arcade/blob/05cb24592af3989d8f274e68280a3c16520d2f71/azure-pipelines.yml#L49-L54

dougbu committed:
Move to BYOC queues
- aspnet/AspNetCore-Internal#2033
- use conditions matching aspnet/Extensions and dotnet/Arcade
  - e.g. https://github.com/dotnet/arcade/blob/05cb24592af3989d8f274e68280a3c16520d2f71/azure-pipelines.yml#L49-L54
- install necessary Build Tools for Visual Studio 2019 components and workloads on CI
- revert part of 85ae18c723 because multiple editions of VS 2019 are now publicly available
  - that is, restore support for multiple editions in InstallVisualStudio.ps1
- update InstallVisualStudio.ps1 to
  - support BuildTools edition of VS 2019

nit: merge a couple of Windows build steps
  - support automatically updating latest version of a VS edition  on the machine
  - include `-Quiet` option for a completely non-interactive installation

dougbu commented:
Like aspnet/AspNetCore-Tooling#352, this needs more work

dougbu committed:
Install SQL Server 2017 Express LocalDB in Windows test job

dougbu commented:
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=132367

dougbu commented:
All failure look like flakiness. Not marking any tests as flaky because I believe others have already decided not to do that and I'm not sure why.

@aspnet/brt please see
- aspnet/AspNetCore-Internal#2004 (Windows build)
- aspnet/AspNetCore-Internal#1432 (Windows test)
- aspnet/AspNetCore-Internal#2052 (macOS test)

@ryanbrandenburg is @aspnet-hello not working completely reliably? Or, does it have trouble when multiple jobs in a pipeline fail? Either way, the bot didn't triage as I requested.

dougbu commented:
@ajcvickers 6 tests are failing with the latest agent configuration. The builds use SQL Server 2017 LocalDB 14.0.1000.169. Does that version or one of the tests do something odd with .mdf file placement? The path `C:\Users\vsagentMicrosoft.AspNetCore.Diagnostics.EntityFrameworkCore.FunctionalTests.Scratch_6.mdf` looks mangled.

```
  [xUnit.net 00:00:43.66]     Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests.MigrationsEndPointMiddlewareTest.Migration_request_default_path [FAIL]
  Failed   Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests.MigrationsEndPointMiddlewareTest.Migration_request_default_path
  Error Message:
   System.InvalidOperationException : An error occurred while applying the migrations for 'Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests.BloggingContextWithMigrations'. See InnerException for details.System.Data.SqlClient.SqlException (0x80131904): CREATE FILE encountered operating system error 5(Access is denied.) while attempting to open or create the physical file 'C:\Users\vsagentMicrosoft.AspNetCore.Diagnostics.EntityFrameworkCore.FunctionalTests.Scratch_6.mdf'.
  CREATE DATABASE failed. Some file names listed could not be created. Check related errors.
     at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
     at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
     at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
     at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
     at System.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds(String methodName, Boolean async, Int32 timeout, Boolean asyncWrite)
     at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, Boolean sendToPipe, Int32 timeout, Boolean asyncWrite, String methodName)
     at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
     at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
     at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
     at Microsoft.EntityFrameworkCore.Migrations.MigrationCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
     at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IRelationalConnection connection)
     at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerDatabaseCreator.Create()
     at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
     at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.Migrate(DatabaseFacade databaseFacade)
     at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.MigrationsEndPointMiddleware.Invoke(HttpContext context) in /_/src/Middleware/Diagnostics.EntityFrameworkCore/src/MigrationsEndPointMiddleware.cs:line 82
  ClientConnectionId:cf288411-eca9-4f4a-8553-a8f7b2228b10
  Error Number:5123,State:1,Class:16
  ---- System.Data.SqlClient.SqlException : CREATE FILE encountered operating system error 5(Access is denied.) while attempting to open or create the physical file 'C:\Users\vsagentMicrosoft.AspNetCore.Diagnostics.EntityFrameworkCore.FunctionalTests.Scratch_6.mdf'.
  CREATE DATABASE failed. Some file names listed could not be created. Check related errors.
  Stack Trace:
     at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.MigrationsEndPointMiddleware.Invoke(HttpContext context) in /_/src/Middleware/Diagnostics.EntityFrameworkCore/src/MigrationsEndPointMiddleware.cs:line 94
     at Microsoft.AspNetCore.TestHost.HttpContextBuilder.<>c__DisplayClass14_0.<<SendAsync>b__0>d.MoveNext() in /_/src/Hosting/TestHost/src/HttpContextBuilder.cs:line 74
  --- End of stack trace from previous location where exception was thrown ---
     at Microsoft.AspNetCore.TestHost.ClientHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /_/src/Hosting/TestHost/src/ClientHandler.cs:line 120
     at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
     at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests.MigrationsEndPointMiddlewareTest.Migration_request(Boolean useCustomPath) in /_/src/Middleware/Diagnostics.EntityFrameworkCore/test/FunctionalTests/MigrationsEndPointMiddlewareTest.cs:line 112
     at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests.MigrationsEndPointMiddlewareTest.Migration_request_default_path() in /_/src/Middleware/Diagnostics.EntityFrameworkCore/test/FunctionalTests/MigrationsEndPointMiddlewareTest.cs:line 58
  --- End of stack trace from previous location where exception was thrown ---
  ----- Inner Stack Trace -----
     at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
     at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
     at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
     at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
     at System.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds(String methodName, Boolean async, Int32 timeout, Boolean asyncWrite)
     at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, Boolean sendToPipe, Int32 timeout, Boolean asyncWrite, String methodName)
     at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
     at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
     at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
     at Microsoft.EntityFrameworkCore.Migrations.MigrationCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
     at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IRelationalConnection connection)
     at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerDatabaseCreator.Create()
     at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
     at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.Migrate(DatabaseFacade databaseFacade)
     at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.MigrationsEndPointMiddleware.Invoke(HttpContext context) in /_/src/Middleware/Diagnostics.EntityFrameworkCore/src/MigrationsEndPointMiddleware.cs:line 82
```

aspnet-hello commented:
This comment was made automatically. If there is a problem contact ryanbrandenburg.

I've triaged the above build.

dougbu commented:
@ajcvickers thanks for the comment in aspnet/EntityFrameworkCore#10288

Think I'll switch to use the SQL Server 101**6** Express installer since that's what https://www.microsoft.com/en-us/sql-server/sql-server-editions-express points to for LocalDB today (despite being a page about the  1017 release) and the cumulative update is twice the size of the Express installer but might not work with Express editions. (Can't find any updates for the Express editions.)

dougbu committed:
Drop back to SQL Server 2016 Express LocalDB

ajcvickers commented:
@dougbu Fair enough.

ErikEJ commented:
@dougbu Direct download links are here: https://stackoverflow.com/questions/42951632/sql-server-express-localdb-msi-offline-installer/42952186#42952186

ErikEJ commented:
The SQL Server CU updates any edition, include Express and "Express LocalDB"

dougbu commented:
Thanks for the direct link and the CU information. Downloading the MSI directly will certainly simplify my installation script. I'll bounce back to 2017 in the next iteration.

ErikEJ commented:
You are welcome, yes, I had a look a that script, it looked a bit hairy.

ErikEJ commented:
Feel free to ping me for any help!

ajcvickers commented:
Thanks Erik. You're always on top of the SQL Server download foo. :-)

dougbu committed:
Instead: Install SQL Server 2017 Express LocalDB and its cumulative update

dougbu committed:
Don't move Microsoft.AspNetCore.Blazor.Templates.dll into output directory
- assembly is not of interest and should not be signed

SteveSandersonMS commented:
The change under `Components` looks harmless to me.

dougbu commented:
Reopening to see if that breaks the Linux jobs out of jail.

dougbu committed:
Revert most of "Don't move Microsoft.AspNetCore.Blazor.Templates.dll into output directory"
- didn't help odd attempt to sign this file and it was only a warning

This reverts commit b55d69c3709f2fdfc460d7de9e18d4b18a7a79b2.

dougbu committed:
Back to using SQL Server 2016 Express LocalDB again
- installing cumulative update was taking extra time and not reliably updating the LocalDB install
- reverts part of 70d8d125f9, leaving direct download and some reordering

dougbu commented:
:up::date: to use SQL Server 2016 LocalDB again
- the only failures in the last check were evidence the cumulative update didn't update LocalDB

NTaylorMullen commented on eng/scripts/InstallVisualStudio.ps1:
```diff
@@ -54,35 +48,57 @@ mkdir $intermedateDir -ErrorAction Ignore | Out-Null
 
-& $bootstrapper @arguments
+$process = Start-Process -FilePath "$bootstrapper" -ArgumentList $arguments `
+    -PassThru -RedirectStandardError "$intermedateDir\errors.txt" -Verbose -Wait
```
Interesting, so if you don't redirect it to a file to then write it out it doesn't make it to the console output? Also, if you're taking this approach probably smart to remove the errors.txt before running this in case someone cancelled the build.

dougbu replied:
The file is removed just before exiting this script.

NTaylorMullen commented on .azure/pipelines/jobs/default-build.yml:
```diff
@@ -115,20 +120,25 @@ jobs:
-    - powershell: ./eng/scripts/InstallJdk.ps1 '11.0.1'
+    - powershell: |
+        ./eng/scripts/InstallJdk.ps1 '11.0.1'
+        Write-Host "##vso[task.prependpath]$env:JAVA_HOME\bin"
```
Not quite sure the reasoning for moving all of this around. Seems like it's less verbose now.

dougbu replied:
Just cleaning up some duplication and shortening this build. The `|` combines the following lines into a single build step.

NTaylorMullen reviewed (approved):
Did my best to review but I lack a lot of context on each of the pieces of this PR. from what I gather you're:
- Installing local DB prior to the primary job
- Changing the build queues for public/internal PRs
- Outputting some extra information about the java home dir on the agent
- Installing Chrome so selenium can run I take it?
- Updated the VS installer to require 16.0 VS and added a quiet flag
- Added a load of dependencies in order to properly build the repo.

dougbu commented:
Thanks for the review @NTaylorMullen!

A few clarifications / answers:
> - Installing local DB prior to the primary job

Correct but only in the "Test: Windows &hellip;" job. The default-build.yml file defines the template used in all jobs. The ci.yml file uses this template, defining the actual builds i.e. passing information into the template.
> - Changing the build queues for public/internal PRs

Yes. This was the _intent_ of the change. Most of the rest is in reaction to these new agents having the BuildTools edition of VS (rather than Enterprise) installed.
> - Outputting some extra information about the java home dir on the agent
> - Installing Chrome so selenium can run I take it?

No, just combining build steps to make the YAML and the build output less noisy.
> - Updated the VS installer to require 16.0 VS and added a quiet flag

Yes to adding the `-Quiet` flag but the script installed VS 2019 before. Main change was to restore support for multiple editions now that they're publicly available.
> - Added a load of dependencies in order to properly build the repo.

Sort of. The new dependencies file is for the BuildTools installation. Workload names are different (mostly adding "BuildTools") because the "full" ones aren't available in this edition. Also needed to add a couple of components that are optional in this edition.

dougbu committed:
Do not assume vsjitdebuger.exe exists
- component is not available in BuildTools edition of VS
